src/eval/mod.rs

```diff
@@ -762,6 +767,12 @@ fn update_thunks(stack: &mut Stack, closure: &Closure) {
     }
 }
 
+fn reset_thunks(stack: &mut Stack) {
+    while let Some(thunk) = stack.pop_thunk() {
+        thunk.reset_state();
+    }
+}
```

Comment:
I fear that won't be enough. The stack could very much look like:

```
ThunkUpdateFrame
Op2Cont
...
ThunkUpdateFrame
```

The second thunk won't be released. I think you have to iterate through all the stack until the end, and treat thunk update frames, and not only the first consecutive thunks.

Reply:
Hmm true, I somehow overlooked this

Reply:
Should work correctly now